kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java

```diff
@@ -58,16 +56,51 @@ public void initChannel(SocketChannel ch) {
 
         ChannelPipeline pipeline = ch.pipeline();
 
-        sslContext.ifPresent(s -> {
+        int targetPort = ch.localAddress().getPort();
+        if (requireTls) {
             LOGGER.debug("Adding SSL/SNI handler");
-            pipeline.addLast(new SniHandler(new DomainWildcardMappingBuilder<>(s).build()));
-        });
+            pipeline.addLast(new SniHandler(hostname -> {
+                var virtualCluster = virtualClusterResolver.resolve(hostname, targetPort);
+                // TODO Error handling.
+                var sslContext = virtualCluster.buildSslContext();
+                if (sslContext.isEmpty()) {
+                    throw new IllegalStateException("Virtual cluster %s does not provide SSL context".formatted(virtualCluster));
+                }
+                ch.attr(VIRTUAL_CLUSTER).set(virtualCluster);
+                return sslContext.get();
+            }));
+
+            // Intercept the SniCompletionEvent and use that as a trigger to add the remainder of the pipeline.
+            pipeline.addLast(new ChannelInboundHandlerAdapter() {
+                @Override
+                public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+                    // The order is important here, we need to add the handlers before propagating the user event.
+                    // This is important as our pipeline line needs to hear the SNI event.
+                    if (evt instanceof SniCompletionEvent) {
+                        var virtualCluster = ((VirtualCluster) ctx.channel().attr(VIRTUAL_CLUSTER).get());
+                        addHandlers(ch, virtualCluster);
+                        ctx.fireChannelActive();
+                        ctx.pipeline().remove(this);
+                        ctx.channel().attr(VIRTUAL_CLUSTER).set(null);
+                    }
+                    super.userEventTriggered(ctx, evt);
+                }
+            });
+        }
+        else {
+            var virtualCluster = virtualClusterResolver.resolve(targetPort);
+            addHandlers(ch, virtualCluster);
+        }
+    }
 
-        if (logNetwork) {
+    private void addHandlers(SocketChannel ch, VirtualCluster virtualCluster) {
+        ChannelPipeline pipeline = ch.pipeline();
+        if (virtualCluster.isLogNetwork()) {
             pipeline.addLast("networkLogger", new LoggingHandler("io.kroxylicious.proxy.internal.DownstreamNetworkLogger", LogLevel.INFO));
         }
 
         // Add handler here
+        // TODO FIXME this is in the wrong place, proxy protocol comes over the wire first (before SSL). KW broke this when SSL was introduced.
```

Comment:
This was already broken (by me).  ha proxy protocol takes place before the TLS handshake (before the SNI is known). Maybe this needs to be a server wide property.

Reply:
Raised as separate issue https://github.com/kroxylicious/kroxylicious/issues/287. Let's fix this later.